src/core/raster/qgsrasterlayerrenderer.cpp

```diff
@@ -273,7 +276,14 @@ QgsRasterLayerRenderer::QgsRasterLayerRenderer( QgsRasterLayer *layer, QgsRender
        && !( rendererContext.flags() & Qgis::RenderContextFlag::RenderPreviewJob )
        && !( rendererContext.flags() & Qgis::RenderContextFlag::Render3DMap ) )
   {
-    layer->refreshRendererIfNeeded( rasterRenderer, rendererContext.extent() );
+    double min = std::numeric_limits<double>::quiet_NaN();
+    double max = std::numeric_limits<double>::quiet_NaN();
+    if ( layer->refreshRendererIfNeeded( rasterRenderer, rendererContext.extent(), min, max ) )
```

Comment:
I am wondering if we can do better now that you've got access to the statistics in QgisApp::onCalculatedRenderDetailsChanged. Modifying the layer directly here is really not a very nice design. I think a cleaner approach would be to:

1. Move out the logic in QgsRasterLayer::refreshRenderer which calculates the min/max for the extent to a separate function, and use that here to retrieve the min/max.
2. Move the remaining logic from QgsRasterLayer::refreshRendererIfNeeded to a function in QgsRasterRenderer, and then here call that method. 

This way we won't be doing ANY modifications to the layer itself from QgsRasterLayerRenderer, which is how the QgsMapLayerRenderer subclasses are supposed to behave.

Reply:
Thanks for the suggestion. However, I'm afraid I don't fully understand it. At the moment, there is already a function in `QgsRasterLayer` which computes the Min/Max: `computeMinMax`.   
`QgsRasterLayer::refreshRenderer` does 2 things: 
- refresh the renderer associated with the LayerRenderer by calling computeMinMax and setClassificationMin/Max for each band
- refresh the layers's renderer with the same logic

I tried to implement a quick POC based on my understanding: 
- I have added `QgsRasterRenderer::needsRefresh` which checks if the renderer needs to be updated
- I have added `QgsRasterRenderer::refresh` overriden by each renderer which basically calls setClassificationMin/Max based on Min/Max values already computed

Now, `QgsRasterLayerRenderer` calls `QgsRasterRenderer::needsRefresh` on its rendered. If true:
- Min/Max are computed and stored in `QgsRenderedLayerStatistics`
- the renderer of `QgsRasterLayerRenderer` is refresh with some Min/Max values
- In `QgisApp` `QgsRasterRenderer::refresh` is called on the layer renderer
- if the modified canvas is the main one, the style and legends are updated

Does this match what you had in mind?

Reply:
Yes, that sounds right! My main issue with the current setup is how QgsRasterLayer::refreshRenderer changes the layer's renderer, and your approach avoids this. :+1: